File: training-data/csharp/InventorySystem.cs
using System;
using System.Collections.Generic;
using System.Linq;

sealed record Product(string Sku, string Name, decimal Price, int ReorderLevel)
{
    public int Quantity { get; set; }
}

sealed class Inventory
{
    readonly Dictionary<string, Product> products = new();

    public void Add(Product p) => products.Add(p.Sku, p);

    Product Get(string sku) => products.TryGetValue(sku, out var p) ? p : throw new KeyNotFoundException($"unknown sku {sku}");

    public void Receive(string sku, int qty) => Get(sku).Quantity += qty;

    public decimal Sell(string sku, int qty)
    {
        var p = Get(sku);
        if (p.Quantity < qty) throw new InvalidOperationException($"{p.Name}: have {p.Quantity}, want {qty}");
        p.Quantity -= qty;
        return qty * p.Price;
    }

    public IEnumerable<string> ReorderList() => products.Values.Where(p => p.Quantity <= p.ReorderLevel).Select(p => p.Name);
    public decimal Valuation => products.Values.Sum(p => p.Quantity * p.Price);
}

class InventorySystem
{
    static void Main()
    {
        var inv = new Inventory();
        inv.Add(new Product("A1", "apple", 0.50m, 10));
        inv.Add(new Product("P1", "pear", 0.75m, 5));
        inv.Receive("A1", 40);
        inv.Receive("P1", 8);
        decimal revenue = inv.Sell("A1", 32) + inv.Sell("P1", 2);
        try
        {
            inv.Sell("P1", 100);
        }
        catch (InvalidOperationException e)
        {
            Console.WriteLine($"error: {e.Message}");
        }
        Console.WriteLine($"revenue {revenue:F2}, stock value {inv.Valuation:F2}, reorder: {string.Join(",", inv.ReorderList())}");
    }
}

File: training-data/ruby/inventory_system.rb
class Inventory
  class UnknownSku < StandardError; end
  class OutOfStock < StandardError; end

  Product = Struct.new(:sku, :name, :price_cents, :reorder_level, :qty, keyword_init: true)

  def initialize = @products = {}

  def add(sku:, name:, price_cents:, reorder_level: 5)
    @products[sku] = Product.new(sku: sku, name: name, price_cents: price_cents, reorder_level: reorder_level, qty: 0)
  end

  def receive(sku, qty) = fetch(sku).qty += qty

  def sell(sku, qty)
    product = fetch(sku)
    raise OutOfStock, "#{product.name}: have #{product.qty}, want #{qty}" if product.qty < qty

    product.qty -= qty
    qty * product.price_cents
  end

  def reorder_list = @products.values.select { |p| p.qty <= p.reorder_level }.map(&:name)
  def valuation_cents = @products.values.sum { |p| p.qty * p.price_cents }

  private

  def fetch(sku) = @products.fetch(sku) { raise UnknownSku, sku }
end

inv = Inventory.new
inv.add(sku: 'A1', name: 'apple', price_cents: 50, reorder_level: 10)
inv.add(sku: 'P1', name: 'pear', price_cents: 75)
inv.receive('A1', 40)
inv.receive('P1', 8)
revenue = inv.sell('A1', 32) + inv.sell('P1', 2)
begin
  inv.sell('P1', 100)
rescue Inventory::OutOfStock => e
  puts "error: #{e.message}"
end
puts format('revenue %.2f, stock value %.2f, reorder %s', revenue / 100.0, inv.valuation_cents / 100.0, inv.reorder_list.inspect)
